### PR TITLE
Move configuration objects into config module

### LINKED ISSUE #5214: Redesign config module
Refactor config parsing and initialization such that we have a module `ert.config` which exports `ErtConfig`, `EnkfObs`, and `ConfigValidationError`. It should have a private sub module `ert.config.parser` which contains the lark parser. The public interface for reading config files should be through `ErtConfig.from_file` and `EnkfObs.from_file`.

### PR DESCRIPTION
**Pull request by eivindjahren**
**Issue**
Resolves #5214

## Pre review checklist

- [x] Added appropriate release note label
- [x] PR title captures the intent of the changes, and is fitting for release notes.
- [x] Commit history is consistent and clean, in line with the [contribution guidelines](https://github.com/equinor/ert/blob/main/CONTRIBUTING.md).
- [x] Updated documentation
- [x] Ensured new behaviour is tested

Adding labels helps the maintainers when writing release notes. This is the [list of release note labels](https://github.com/equinor/ert/labels?q=release-notes).

### CODE REVIEW

**Review by yngve-sk — APPROVED** [review]
Loosk good, moved some stuff, changed some imports🚀

File: tests/unit_tests/c_wrappers/res/enkf/test_ert_config.py [hunk 1]
```diff
@@ -186,10 +190,6 @@ def test_extensive_config(setup_case):
         snake_oil_structure_config["FORWARD_MODEL"]
         == ert_config.forward_model_job_name_list()
     )
-    assert (
-        HistorySource[snake_oil_structure_config["HISTORY_SOURCE"]]
-        == model_config.history_source
-    )
```
Comment (andreas-el) [inline tests/unit_tests/c_wrappers/res/enkf/test_ert_config.py:192]: Just out of curiosity, this was removed on purpose? Maybe not applicable anymore?

**Review by andreas-el — APPROVED** [review]
Just the one question

File: tests/unit_tests/c_wrappers/res/enkf/test_ert_config.py (hunk 1 above)
Comment (eivindjahren) [reply tests/unit_tests/c_wrappers/res/enkf/test_ert_config.py:192]: Yes, removed on purpose as we test that in other ways.

**Pull request merged**